pkg/eventqueue/eventqueue.go

```diff
@@ -0,0 +1,207 @@
+// Copyright 2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package eventqueue
+
+import (
+	"sync"
+
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "eventqueue")
+)
+
+// EventQueue is a structure which is utilized to handle Events in a first-in,
+// first-out order.There are several important properties for an EventQueue. An
+// EventQueue may be implemented in a blocking manner, such that only one Event
+// is able to be consumed at a time per event queue. This is useful in the case
+// for ensuring that certain types of events do not run concurrently. An
+// EventQueue may be closed, in which case all events which are queued up, but
+// have not been processed yet, will be cancelled (i.e., not ran). It is
+// guaranteed that no events will be scheduled onto an EventQueue after it has
+// been closed; if any event is attempted to be scheduled onto an EventQueue
+// after it has been closed, it will be cancelled immediately. For any event to
+// be processed by the EventQueue, it must implement the `EventHandler`
+// interface. This allows for different types of events to be processed by
+// anything which chooses to utilize an `EventQueue`.
+type EventQueue struct {
+	// This should always be a buffered channel.
+	events chan *Event
+	// close is closed once the EventQueue has been closed.
+	close chan struct{}
+	// drained is closed once the events channel is drained.
+	drained chan struct{}
+	// eventQueueOnce is used to ensure that the EventQueue business logic can
+	// only be ran once.
+	eventQueueOnce sync.Once
+}
+
+// NewEventQueue returns an EventQueue with a capacity for only one event at
+// a time, and all other needed fields initialized.
+func NewEventQueue() *EventQueue {
+	return &EventQueue{
+		// Only one event can be consumed at a time.
+		events:  make(chan *Event, 1),
+		close:   make(chan struct{}),
+		drained: make(chan struct{}),
+	}
+
+}
+
+// Event is an event that can be enqueued onto an EventQueue.
+type Event struct {
+	// Metadata is the information about the event which is sent
+	// by its queuer. Metadata must implement the EventHandler interface in
+	// order for the Event to be successfully processed by the EventQueue.
+	Metadata interface{}
```

Comment:
This would allow removal of the weird `e.Metadata.(type)` switch below:

```suggestion
	Metadata EventHandler
```

(And the comment could be updated, because there's no point stating that it's an EventHandler if the golang code defines it this way).

Reply:
Done!